evolver/history/interface.py

```diff
@@ -0,0 +1,54 @@
+from abc import abstractmethod
+from typing import Any
+
+from pydantic import BaseModel
+
+from evolver.base import BaseInterface
+
+
+class HistoricDatum(BaseModel):
+    timestamp: float
+    data: Any
+
+
+class HistoryResult(BaseModel):
+    data: dict[str, list[HistoricDatum]]
+
+
+class History(BaseInterface):
+    class Config(BaseInterface.Config):
+        pass
+
+    @abstractmethod
+    def put(self, name: str, data: Any):
+        """Add data for hardware component to history.
+
+        Args:
+            name: The name of hardware component to save data for.
+            data: The data to save. This MUST be a JSON-serializable object, as
+                per the FastAPI json serialization requirements.
+        """
+        pass
+
+    @abstractmethod
+    def get(self, name: str = None, t_start: float = None, t_stop: float = None, n_max: int = None) -> HistoryResult:
+        """Get saved history data.
```

Comment:
Interesting, the way it is stored now is that a history entry is a single "read out" of the hardware, which contains all the data for all vials in a single point. so for example:

```
time, name, output
1234, "od", {"0": {'vial': 0, 'raw': 1234}, "1": {'vial': 1, 'raw': 1234}}
```

which would come out in history as:
```json
{ "data": { 
  "od": [ 
    {"timestamp": 1234, 
      "data": {"0": {"vial": 0, "raw": 1234}, "1": {"vial": 1, "raw": 1234}}
     }
   ]
}}
```

so a filter on vial or properties would effectively be pruning the data in output column (generally we would expect a read on all available vials in each timepoint).

would the goal be primarily to reduce result size?

Reply:
Goal is selfishly to support views on the front end, for historical data filtered by vials and properties. e.g. Show me vials 1 and 2, temp readings, historical. These are implemented but there's a lot of mapping and filtering to collate the data...

Could `json_extract` let us declaratively cherry pick data from the 3rd `output` column's json:
https://duckdb.org/docs/extensions/json.html#json-extraction-functions

Though perhaps this would be a different endpoint and different response shape, can talk about it in person. :)